Monthlyactivity

pamzey opened the pull request:
This pull request optimizes the query and its sub-functions to eliminate looping through patient observations, resulting in improved performance and efficiency. The key changes include
Refactored Query: The SQL query has been refactored to minimize sub-queries and enhance readability.
JOINs Optimization: We've optimized JOIN operations, reducing the need for nested sub-queries.
Indexing: Ensured that relevant columns used in JOIN and WHERE clauses are properly indexed for faster data retrieval.
These changes aim to enhance the overall performance of the query and make it more efficient for processing large datasets.
This fix is for all MoH reports and the perfomance has improved

pamzey committed:
removed dublicates on the monthly activity report

pamzey committed:
removed obs looping

pamzey committed:
Merge branch 'development' of https://github.com/HISMalawi/BHT-EMR-API into monthlyactivity

pamzey committed:
removed dublicated code

pamzey committed:
removed dublicated code

Skipper-116 commented on app/services/hts_service/reports/hts_report_builder.rb:
```diff
@@ -38,115 +38,94 @@ def his_patients_rev
+              INNER JOIN obs ON obs.person_id = patient.patient_id
+              WHERE
+                patient.voided = 0
+                AND encounter.encounter_datetime BETWEEN '#{@start_date}' AND '#{@end_date}'
```
You need to add these filters on join themselves to avoid a full table scan. Especially you are using inner joins that will narrow down the data

Skipper-116 commented on app/services/hts_service/reports/hts_report_builder.rb:
```diff
@@ -38,115 +38,94 @@ def his_patients_rev
+                encounter.encounter_datetime,
+                obs.*
+              FROM patient
+              INNER JOIN person ON person.person_id = patient.patient_id
```
also handle voided records here

Skipper-116 commented on app/services/hts_service/reports/hts_report_builder.rb:
```diff
@@ -38,115 +38,94 @@ def his_patients_rev
+                obs.*
+              FROM patient
+              INNER JOIN person ON person.person_id = patient.patient_id
+              INNER JOIN encounter ON encounter.patient_id = patient.patient_id
```
```suggestion
INNER JOIN encounter ON encounter.patient_id = patient.patient_id AND encounter.voided = 0
```
the filters in the where clause should also come here

Skipper-116 commented on app/services/hts_service/reports/hts_report_builder.rb:
```diff
@@ -38,115 +38,94 @@ def his_patients_rev
+              FROM patient
+              INNER JOIN person ON person.person_id = patient.patient_id
+              INNER JOIN encounter ON encounter.patient_id = patient.patient_id
+              INNER JOIN program ON program.program_id = encounter.program_id
```
same here add voided filters for program i believe it should be retired

Skipper-116 commented on app/services/hts_service/reports/hts_report_builder.rb:
```diff
@@ -38,115 +38,94 @@ def his_patients_rev
+              INNER JOIN person ON person.person_id = patient.patient_id
+              INNER JOIN encounter ON encounter.patient_id = patient.patient_id
+              INNER JOIN program ON program.program_id = encounter.program_id
+              INNER JOIN obs ON obs.person_id = patient.patient_id
```
Especially here
```
AND voided = 0
```

Skipper-116 commented on app/services/hts_service/reports/hts_report_builder.rb:
```diff
@@ -38,115 +38,94 @@ def his_patients_rev
+              desired_observation = observations.find { |obs| obs["concept_id"] == indicator[:concept_id] }
+              if desired_observation.present?              
+                val = indicator[:value]
+                name = indicator[:name]
```
Just curious how does the front-end handle response of indicators with indexes and those without?

For example line 86 and line 97 are different

pamzey committed:
added filters to the join

pamzey commented:
I have made  necessary changes as you suggested and about how the front-end handles indicators with and without indexes, both lines 86 and 97 in the code are identical its just that one handles retrieving of data in array and the other with data not in array the front-end treats indicators with and without indexes in the same way. In both cases, it checks if the indicator is of type value_numeric  and converts the value accordingly. If not, it directly assigns the value. The front-end can then use these values as needed for data visualization or other purposes.

Skipper-116 commented on app/services/hts_service/reports/hts_report_builder.rb:
```diff
@@ -38,115 +38,95 @@ def his_patients_rev
-            indicators.each do |indicator|
-              if indicator[:concept_id].is_a?(Integer)
-                desired_observation = observations.find { |obs| obs["concept_id"] == indicator[:concept_id].to_s }
+          indicators.each do |indicator|
```
Can you replace this loop with this
```suggestion
indicators.each do |indicator|
    indicator_array = indicator[:concept_id].is_a?(Array) ? indicator[:concept_id] : [indicator[:concept_id]]
    indicator_values = indicator[:value]
    indicator_names = indicator[:name]
    indicator_array.each_with_index do |concept_id, index|
        desired_observation = observations.find { |obs| obs['concept_id'] == concept_id }
        name = indicator_names[index]
        data[index_new][name] = desired_observation&.[](indicator_values) if desired_observation
    end
end
```

pamzey committed:
Merge branch 'development' of https://github.com/HISMalawi/BHT-EMR-API into monthlyactivity

pamzey committed:
Refactor indicator processing loop for conciseness

pamzey commented:
I have refactored indicator processing loop for clarity and conciseness from your guide and also i have updated the sub queries to avoid searching the whole tables which was causing the query to be slow.

Skipper-116 commented:
@pamzey  this looks okay, i have added @andie23  for another eye on this
